.travis.yml

```
@@ -1,42 +1,39 @@
-﻿dist: trusty
-sudo: false
-env:
-  global:
-    - MAKEFLAGS="-j 2"
+﻿os: linux
+dist: bionic
 
 language: perl
+# See https://www.cpan.org/src/
 perl:
   - 5.30
   - 5.28
   - 5.26
   - 5.24
   - 5.22
   - 5.20
-  - 5.18
-  - 5.16
-  - 5.14
-  - 5.12
-  - 5.10
-  - 5.8
 
 cache:
   directories:
-  - ~/perl_modules
-
+    - ~/perl_modules
 before_install:
+  - export MAKEFLAGS="-j $(nproc)"
   - cpanm --quiet --notest local::lib
   - eval "$(perl -Mlocal::lib=~/perl_modules)"
-
 install:
   - cpanm --quiet --installdeps --notest .
-
 before_script:
   - ./makeDist --nosyntaxCheck --version HEAD
   - cd ./dist/BackupPC-HEAD
   - sudo useradd -s /usr/sbin/nologin -G travis backuppc
-
 script:
-  - sudo PATH="$PATH" PERL5LIB="$PERL5LIB" ./configure.pl --batch
-  - sudo PATH="$PATH" PERL5LIB="$PERL5LIB" su -m backuppc -c '/usr/local/BackupPC/bin/BackupPC -d'
+  - sudo --preserve-env=PATH,PERL5LIB ./configure.pl --batch
+  - sudo --preserve-env=PATH,PERL5LIB su -m backuppc -c '/usr/local/BackupPC/bin/BackupPC -d'
   - sudo cat /var/log/BackupPC/LOG
   - sudo egrep -v '[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2} (Reading hosts file|BackupPC HEAD \(Perl v5\.[0-9]{1,2}\.[0-9]\) started, pid [0-9]+|Next wakeup is [0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2})$' /var/log/BackupPC/LOG; test $? -eq 1
+
+jobs:
+  - stage: lint
+    language: python
+    python: 3.8
+    script:
+      - pip3 install pre-commit
+      - pre-commit run --all --color always
```

This is not working since we don't need to buld and run BackupPC before linting.
```suggestion
  include:
    - name: Lint
      language: python
      python: 3.8
      before_install: skip
      install: skip
      before_script: skip
      script:
        - pip3 install pre-commit
        - pre-commit run --all --color always
```

---

How come ? I tested it on my fork and it was working nicely. Did you setup your own travis ?

---

This is what I get https://travis-ci.org/github/jooola/backuppc/jobs/693813005/config

---

> How come ? I tested it on my fork and it was working nicely. Did you setup your own travis ?
No, I just enabled it for my own fork.
As you can see the test failed for this PR as well: https://travis-ci.org/github/backuppc/backuppc/jobs/693814175

---

Ok, yes, see why I haven't noticed this.

---

I still don't understand why do you want to run lint tests in a separate stage? They are independent from other tests, why not run them in parallel?
Moreover, if one or more of the jobs of `test` (default) stage has failed, `lint` stage will not start at all, but lint result can be useful even in this case. Let assume a developer made an error in one line that breaks build on some Perl version and at the same time they broke code style in the other line.

---

@moisseev This is the first time I really use travis, I used to play around for some basic things, but I uncertain on know how to properly set this up.

I might have misunderstood how it works. What you say seem better anyway, I introduced an extra stage to prevent the before_* to be executed, but it seems I was totally wrong.

---

I am not an expert either. Stages are described [here](https://docs.travis-ci.com/user/build-stages#what-are-build-stages). Another thing, the jobs under `jobs: include:` inherit everything from default job settings. I am not sure if it is elegant or ideologically correct, but the solution that I proposed in the suggestion above is tested and working.

---

Ok, I definitely dislike the travis's default job and unconditionally inheriting from the default job.
But I understand now how it works !

Anyway, I changed the setup.

---

It works, thanks!